add structured suggestions and fix false-positive for elided-lifetimes-in-paths lint

zackmdavis opened the pull request:
This adds structured suggestions to the elided-lifetimes-in-paths lint (introduced in Nov. 2017's #46254), prevents it from emitting a false-positive on anonymous (underscore) lifetimes (!), and adds it to the idioms-2018 group (#52041).

~~As an aside, "elided-lifetimes-in-paths" seems like an unfortunate name, because it's not clear exactly what "elided" means. The motivation for this lint (see original issue #45992, and [RFC 2115](https://github.com/rust-lang/rfcs/blob/e978a8d3017a01d632f916140c98802505cd1324/text/2115-argument-lifetimes.md#motivation)) seems to be specifically about not supplying angle-bracketed lifetime arguments to non-`&` types, but (1) the phrase "lifetime elision" has historically also referred to the ability to not supply a lifetime name to `&` references, and (2) an `is_elided` method in the HIR returns true for anoymous/underscore lifetimes, which is _not_ what we're trying to lint here. (That naming confusion is almost certainly what led to the false positive addressed here.) Given that the lint is relatively new and is allow-by-default, is it too late to rename it ... um, _again_ (#50879)?~~

~~This does _not_ address a couple of other false positives discovered in https://github.com/rust-lang/rust/issues/52041#issuecomment-402547901.~~

![elided_states](https://user-images.githubusercontent.com/1076988/42302137-2bf9479c-7fce-11e8-8bd0-f29aefc802b6.png)

r? @nikomatsakis 
cc @nrc @petrochenkov

zackmdavis commented:
(Some would argue that we shouldn't give the lint the increased prominence of the idioms-2018 group (in contrast to allow-by-default, which I fear almost no one turns up) until we fix the false-positive exemplified by `Ref::map` (again, in https://github.com/rust-lang/rust/issues/52041#issuecomment-402547901 ), but the counterargument would be that anyone who enables the idioms-2018 lints before October should know that they're opting into alpha/beta software and might help us discover more edge cases?)

nikomatsakis commented:
@zackmdavis 

I'm good with another name, do you have a suggestion? When it comes to the term "elided", there are two schools of thought in any case:

- "Elided" carries the plain English meaning, and hence refers to any lifetime is not written (my personal opinion)
- "Elided" refers speciically to a lifetime in a return type that is not written and is resolved to a lifetime from one of the parameters ("lifetime elision RFC")

Perhaps best to find another term altogether that is not as contentious. =)

It's sort of unclear what to call a `'_` lifetime, which is not explicitly written but also not invisible... ah, perhaps that is a good name?

`invisible_lifetimes_in_paths`?

nikomatsakis commented:
> This does not address a couple of other false positives

Which ones are not addressed? It'd be nice to update the issue to have a current list of known false positives.

nikomatsakis commented on src/librustc/lint/builtin.rs:
```diff
@@ -252,7 +252,7 @@ declare_lint! {
     pub ELIDED_LIFETIMES_IN_PATHS,
     Allow,
-    "hidden lifetime parameters are deprecated, try `Foo<'_>`"
+    "implicit lifetime parameters are deprecated"
```
As far as names go, my take:

- `implicit` is *ok*, but it tends to mean anything that the reader wants it to mean
- `hidden` and `invisible` I think are both ok

I don't think any name will necessarily convey perfectly what we are going for -- but all 3 names are roughly in the right direction. I would summarize as "a lifetime that you can't tell is there from looking at the type". For this, I like "hidden" or "invisible" better than "implicit".

nikomatsakis commented on src/librustc/middle/resolve_lifetime.rs:
```diff
@@ -2066,27 +2070,78 @@ impl<'a, 'tcx> LifetimeContext<'a, 'tcx> {
+            &format!("implicit lifetime parameters in types are deprecated"),
+        );
+
+        if num_implicit_lifetimes == 1 {
```
Why only the case of 1 implicit lifetime?

nikomatsakis replied:
It seems like the logic below would apply to any number of lifetimes, if we just generalized the string.

nikomatsakis commented on src/librustc/middle/resolve_lifetime.rs:
```diff
@@ -2066,27 +2070,78 @@ impl<'a, 'tcx> LifetimeContext<'a, 'tcx> {
+                    (segment_ident.span.shrink_to_hi(), "<'_>")
+                } else {
+                    // Otherwise—sorry, this is kind of gross—we need to infer the
+                    // replacement point span from the generics that do exist
```
I don't see a better way, but I get nervous around macros and things -- e.g., what will happen if somebody does:

`Ref<$x>`

in a macro? (I'm imagining that the span of `$x` might then come from some other place than the type itself...?) Maybe we want a test around that case? I think it'd be fine to disable suggestions there, but it'd be nice to ICE or suggest something mangled?

nikomatsakis commented on src/librustc/middle/resolve_lifetime.rs:
```diff
@@ -2066,27 +2070,78 @@ impl<'a, 'tcx> LifetimeContext<'a, 'tcx> {
+                    let mut first_generic_span = None;
+                    for ref arg in &generic_args.args {
+                        match arg {
+                            hir::GenericArg::Lifetime(lt) => {
```
I think there cannot be a lifetime name here... you can **totally** elide lifetimes (e.g., `Foo<T>`) but you cannot write only **some** of the needed lifetimes (e.g., `Foo<'_, T>` when `Foo<'_, '_, T>` is needed) -- you'll get an error in the latter case and (probably?) not go down this path. Should test though.

nikomatsakis commented on src/test/ui/in-band-lifetimes/elided-lifetimes.rs:
```diff
@@ -0,0 +1,45 @@
+#![deny(elided_lifetimes_in_paths)]
+//~^ NOTE lint level defined here
+
+use std::cell::{RefCell, Ref};
```
Can we test:

- [ ] some case with more than one parameter (just for fun, though I don't see why it wouldn't work)
- [ ] a case where there is more than one lifetime and you only supply *some*?
- [ ] the macro case where the type is constructed in a macro from bits and pieces?

bors commented:
:umbrella: The latest upstream changes (presumably #52275) made this pull request unmergeable. Please resolve the merge conflicts.

zackmdavis commented:
@nikomatsakis 

(When you're [back from vacation](https://internals.rust-lang.org/t/nikos-availability-over-july-and-august/7904).)

> Why only the case of 1 implicit lifetime? It seems like the logic below would apply to any number of lifetimes, if we just generalized the string.

I couldn't find a situation where inserting `<'_, '_>` would result in successfully fixing broken code. As an illustrative negative example, consider this program—

```rust
#![allow(unused)]
#![warn(elided_lifetimes_in_paths)]

struct DoubleLifetime<'a, 'b> {
    one: &'a str,
    another: &'b str,
}

fn returns_double_lifetime(one: &'a str, another: &'b str) -> DoubleLifetime {
    DoubleLifetime { one, another }
}

fn main() {}
```

The lint does fire on the return type of `returns_double_lifetime`, but we also hit E0106 ("missing lifetime specifiers")—

```
warning: implicit lifetime parameters in types are deprecated
  --> elide.rs:17:63
   |
17 | fn returns_double_lifetime(one: &'a str, another: &'b str) -> DoubleLifetime {
   |                                                               ^^^^^^^^^^^^^^
   |

[...]

error[E0106]: missing lifetime specifiers
  --> elide.rs:17:63
   |
17 | fn returns_double_lifetime(one: &'a str, another: &'b str) -> DoubleLifetime {
   |                                                               ^^^^^^^^^^^^^^ expected 2 lifetime parameters
```

Adding `<'_, '_>` doesn't solve the E0106—

```
error[E0106]: missing lifetime specifiers
  --> elide.rs:17:78
   |
17 | fn returns_double_lifetime(one: &'a str, another: &'b str) -> DoubleLifetime<'_, '_> {
   |                                                                              ^^ expected 2 lifetime parameters
```

I think this makes sense—if there are two distinct "output" lifetimes (in [the terminology of RFC 141](https://github.com/rust-lang/rfcs/blob/master/text/0141-lifetime-elision.md#lifetime-positions)), they can't both be anonymous, because they need names to disambiguate which is which, right? And this is a different scenario from when we can legitimately use multiple anonymous _input_ lifetimes `<'_, '_>` in an impl header [as illustrated in the Edition Guide](https://rust-lang-nursery.github.io/edition-guide/2018/transitioning/ownership-and-lifetimes/anonymous-lifetime.html#more-details)?

> Which ones are not addressed? It'd be nice to update the issue to have a current list of known false positives.

I think I've fixed them now. What _was_ happening is that the lint would fire on the type path in struct construction expressions—

```
error: implicit lifetime parameters in types are deprecated
  --> src/test/ui/in-band-lifetimes/elided-lifetimes.rs:44:5
   |
44 |     WrappedWithBow { gift }
   |     ^^^^^^^^^^^^^^- help: indicate the anonymous lifetime: `<'_>`
```

—and in associated-functions/static-methods (the motivating examples being `Ref::clone` and `Ref::map`)—

```
warning: implicit lifetime parameters in types are deprecated
  --> elide.rs:20:15
   |
20 |     let b2a = Ref::clone(b1);
   |               ^^^-^^^^^^
   |                  |
   |                  help: indicate the anonymous lifetime: `<'_>`
```

In both these cases, applying the suggestion results in code that fails to even parse ("expected `:`, found `>`").

To avoid this, I first moved the lint check from (a callee of) the `visit_path` method in the `LifetimeContext` visitor, to the `hir::TyPath` arm of `visit_ty`. That in itself fixes the struct-constructor case, but we were still emitting a false-positive in `Ref::clone`-like cases. (As far as the HIR is concerned, it seems that `Ref::clone` is the same as the UFCS expression `<Ref>::clone`, where the `Ref` counts as a type to be visited by `visit_ty`.) To fix _that_, I'm afraid I couldn't think of anything better than [just not letting the visitor recurse into the callee of call expressions](https://github.com/rust-lang/rust/commit/4c31c07c4f1b438c5ac48b50ee5202139a1aaf4b). I _think_ that's legitimate, because I'm _pretty sure_ you can't have a named lifetime in need of resolution there?—but **this is in particular need of your expert reviewer scrutiny** before we merge it. ( @oli-obk might also have insight, as the author of [the other work](https://github.com/rust-lang/rust/blame/49f1e5d8ab6edc4249752a0ee53b771d46d8ac49/src/librustc/middle/resolve_lifetime.rs#L628) we're doing in the `TyPath` arm.)

> I think there cannot be a lifetime name here... you can totally elide lifetimes (e.g., Foo<T>) but you cannot write only some of the needed lifetimes (e.g., Foo<'_, T> when Foo<'_, '_, T> is needed) -- you'll get an error in the latter case and (probably?) not go down this path. Should test though.

Thanks for clarifying this. However, unless I can convince myself that we error out _before_ we get here, I think I'm inclined to naïvely obey the type system rather than asserting that the `Lifetime` arm never happens, on separation-of-concerns grounds: this method is _just_ about linting for invisible lifetimes in otherwise (currently!) legal code; erroring out is someone else's responsibility (I added [a comment](https://github.com/zackmdavis/rust/blob/a9fa345dadb18e7cf8dfe6e718bef606fce070f1/src/librustc/middle/resolve_lifetime.rs#L2104-L2106).)

> It's sort of unclear what to call a '_ lifetime, which is not explicitly written but also not invisible...

The [edition guide calls it](https://rust-lang-nursery.github.io/edition-guide/2018/transitioning/ownership-and-lifetimes/anonymous-lifetime.html) an "anonymous" lifetime, which seems appropriate (it's not omitted from the source, it just doesn't have a name). I've followed this terminology in the span label here ("indicate the anonymous lifetime").

> I'm good with another name [...] `invisible_lifetimes_in_paths`? `hidden_lifetime_in_paths`?

After re-consulting [RFC 141](https://github.com/rust-lang/rfcs/blob/master/text/0141-lifetime-elision.md), I'm **not enthusiastic about renaming the lint anymore**.

(The reason I thought we wanted to rename is because I had mentally associated the word "elision" with `&` references specifically, which this lint isn't concerned with, but the RFC makes it clear that unwritten angle-bracked lifetimes are also "elided", and on reflection, I think "in paths" does a good enough job of being more specific.)

rust-highfive commented:
The job `x86_64-gnu-llvm-5.0` of your PR [failed on Travis](https://travis-ci.org/rust-lang/rust/jobs/404051546) ([raw log](https://api.travis-ci.org/v3/job/404051546/log.txt)). Through arcane magic we have determined that the following fragments from the build log may contain information about the problem.

<details><summary><i>Click to expand the log.</i></summary>

```plain
[00:22:21]    Compiling proc_macro v0.0.0 (file:///checkout/src/libproc_macro)
[00:22:38]    Compiling syntax_ext v0.0.0 (file:///checkout/src/libsyntax_ext)
[00:28:39]    Compiling rustc_typeck v0.0.0 (file:///checkout/src/librustc_typeck)
[00:28:39]    Compiling rustc_mir v0.0.0 (file:///checkout/src/librustc_mir)
[00:28:43] error[E0597]: `my_registrar` does not live long enough
[00:28:43]    --> librustc_metadata/creader.rs:582:24
[00:28:43]     |
[00:28:43] 582 |         registrar(&mut my_registrar);
[00:28:43]     |                        ^^^^^^^^^^^^ borrowed value does not live long enough
[00:28:43] 588 |     }
[00:28:43] 588 |     }
[00:28:43]     |     - `my_registrar` dropped here while still borrowed
[00:28:43]     |
[00:28:43]     = note: values in a scope are dropped in the opposite order they are created
[00:28:43] error: variable does not need to be mutable
[00:28:43]    --> librustc_metadata/creader.rs:581:13
[00:28:43]     |
[00:28:43]     |
[00:28:43] 581 |         let mut my_registrar = MyRegistrar { extensions: Vec::new(), edition: root.edition };
[00:28:43]     |             ----^^^^^^^^^^^^
[00:28:43]     |             |
[00:28:43]     |             help: remove this `mut`
[00:28:43]     = note: `-D unused-mut` implied by `-D warnings`
[00:28:43] 
[00:28:43] error: aborting due to 2 previous errors
[00:28:43] 
[00:28:43] 
[00:28:43] For more information about this error, try `rustc --explain E0597`.
[00:28:43] error: Could not compile `rustc_metadata`.
[00:28:43] 
[00:28:43] Caused by:
[00:28:43]   process didn't exit successfully: `/checkout/obj/build/bootstrap/debug/rustc --crate-name rustc_metadata librustc_metadata/lib.rs --color always --error-format json --crate-type dylib --emit=dep-info,link -C prefer-dynamic -C opt-level=2 -C metadata=f2e516612ec0eb19 -C extra-filename=-f2e516612ec0eb19 --out-dir /checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps --target x86_64-unknown-linux-gnu -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps -L dependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/release/deps --extern flate2=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps/libflate2-1a6e055719da9e2b.rlib --extern log=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps/liblog-3aae1b07eee5f41a.rlib --extern proc_macro=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps/libproc_macro-02138f2ab64d4539.so --extern rustc=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps/librustc-da4d2a69d03b2f0d.so --extern rustc_data_structures=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_data_structures-e5f576c4a2daecee.so --extern rustc_errors=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_errors-3575a7cf8f4898f3.so --extern rustc_target=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps/librustc_target-2d5d56ac8c265ecc.so --extern serialize=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps/libserialize-4ea3497d1552c967.so --extern serialize=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps/libserialize-4ea3497d1552c967.rlib --extern syntax=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps/libsyntax-fb9576173c28daa7.so --extern syntax_ext=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps/libsyntax_ext-2d7bddb0d9db098b.so --extern syntax_pos=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/deps/libsyntax_pos-9caed024aacf789e.so -L native=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/build/miniz-sys-00f2106305a03b52/out -L native=/checkout/obj/build/x86_64-unknown-linux-gnu/stage1-rustc/x86_64-unknown-linux-gnu/release/build/backtrace-sys-11ea4ad08c115426/out` (exit code: 101)
travis_time:end:19e9f270:start=1531630179288618005,finish=1531632080027436552,duration=1900738818547

The command "stamp sh -x -c "$RUN_SCRIPT"" exited with 2.
travis_time:start:0f8c0ac0
```

</details><p></p>

[I'm a bot](https://github.com/rust-ops/rust-log-analyzer)! I can only do what humans tell me to, so if this was not helpful or you have suggestions for improvements, please ping or otherwise contact **`@TimNN`**. ([Feature Requests](https://github.com/rust-ops/rust-log-analyzer/issues?q=is%3Aopen+is%3Aissue+label%3Afeature-request))

zackmdavis commented:
(CI error is mysterious; this PR does not touch librustc_metadata)

oli-obk commented:
You definitely broke something, the question is whether this code wasn't broken to begin with:

```rust
        let registrar = unsafe {
            let sym = match lib.symbol(&sym) {
                Ok(f) => f,
                Err(err) => self.sess.span_fatal(span, &err),
            };
            mem::transmute::<*mut u8, fn(&mut dyn Registry)>(sym)
        };
```

zackmdavis commented:
Thanks, my code is almost certainly doing the wrong thing on `mem::transmute::<*mut u8, fn(&mut dyn Registry)>(sym)`. 

I'll need to rethink this (maybe we can store some extra state on the `LifetimeContext` visitor, rather than mucking with its traversal, which I _really_ should have known better than to think I could get away with), but don't have time today.

nikomatsakis commented:
@zackmdavis 

> I couldn't find a situation where inserting <'_, '_> would result in successfully fixing broken code. 

Perhaps this?

```rust
#![allow(unused)]
#![warn(elided_lifetimes_in_paths)]

struct DoubleLifetime<'a, 'b> {
    one: &'a str,
    another: &'b str,
}

fn accepts_double_lifetime(dl: DoubleLifetime) -> String {
    format!("{} {}", dl.one, dl.another)
}

fn main() {}
```

> As far as the HIR is concerned, it seems that `Ref::clone` is the same as the UFCS expression `<Ref>::clone`, where the `Ref` counts as a type to be visited by `visit_ty`.

Hmm, yes, I can imagine that. I wonder if we are doing this lint in the wrong place. Maybe we should do it during **HIR lowering**...? After all, that is where we actually insert the various markers in HIR in the first place. I forget why I didn't suggest that route.

nikomatsakis commented:
> I think that's legitimate, because I'm pretty sure you can't have a named lifetime in need of resolution there?—but this is in particular need of your expert reviewer scrutiny before we merge it.

I would say it's not the worst thing ever, but it seems suboptimal. My goal is roughly to make lifetime parameters work the same as types -- that is, whenever you would have to write the type parameters for `Foo`, you also write the lifetime ones. So if you would do `Vec<_>` (and not just `Vec`) then you should do `Foo<'_>` (and not just `Foo`).

bors commented:
:umbrella: The latest upstream changes (presumably #52264) made this pull request unmergeable. Please resolve the merge conflicts.

zackmdavis commented:
> Maybe we should do it during HIR lowering...?

(This [is looking like](https://github.com/zackmdavis/rust/commit/4f51df67ac61c) it's going to be _much_ better, but for now my build is mysteriously ICEing and I need to sleep.)

zackmdavis commented:
@nikomatsakis ready :checkered_flag:

nikomatsakis commented:
@bors r+ -- looks nice

bors commented:
:pushpin: Commit 6fdcc60d6314d111acf4ea8e4e82cfd20f55f998 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 6fdcc60d6314d111acf4ea8e4e82cfd20f55f998 -->

bors commented:
:hourglass: Testing commit 6fdcc60d6314d111acf4ea8e4e82cfd20f55f998 with merge 2439c83721e73371d2119f1c58c4cb42d8e6f5ea...

bors commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.8224)

kennytm commented:
Caused ICE with `dist-x86_64-msvc-alt` during `BuiltinLintDiagnostics::run()`.

```
   Compiling syn v0.14.4
thread 'main' panicked at 'byte index 89352 is out of bounds of `$crate::synom::PResult<$o>`', libcore\str\mod.rs:2094:9
note: Run with `RUST_BACKTRACE=1` for a backtrace.
```

<details><summary>The stack trace from the debugger:</summary>

```
0000008c`bc9739d8 00007ffd`22a913ed ntdll!NtWaitForMultipleObjects+0xa
0000008c`bc9739e0 00007ffd`23b67d51 KERNELBASE!WaitForMultipleObjectsEx+0xed
0000008c`bc973cc0 00007ffd`23b67773 kernel32!WerpReportFaultInternal+0x581
0000008c`bc974230 00007ffd`22b71c1f kernel32!WerpReportFault+0x83
0000008c`bc974260 00007ffd`258ff1b3 KERNELBASE!UnhandledExceptionFilter+0x23f
0000008c`bc974350 00007ffd`258e1e26 ntdll!RtlUserThreadStart$filt$0+0x3e
0000008c`bc974390 00007ffd`258f349d ntdll!_C_specific_handler+0x96
0000008c`bc974400 00007ffd`258b48d7 ntdll!RtlpExecuteHandlerForException+0xd
0000008c`bc974430 00007ffd`258f262a ntdll!RtlDispatchException+0x197
0000008c`bc974b00 00007ffd`22b6df12 ntdll!KiUserExceptionDispatch+0x3a
*** WARNING: Unable to verify checksum for rustc-158b3752b6ee1be6.dll
0000008c`bc975238 00007ffd`0727c83a KERNELBASE!DebugBreak+0x2
*** WARNING: Unable to verify checksum for std-822da15d609da1ff.dll
0000008c`bc975240 00007ffd`0ca68ca2 rustc_158b3752b6ee1be6!ZN5rustc4util6common10panic_hook17h7a7f3dae1f9492b5E+0x19a
0000008c`bc9752e0 00007ffd`0ca68804 std_822da15d609da1ff!std::panicking::rust_panic_with_hook+0x1a2
0000008c`bc975390 00007ffd`0ca686e9 std_822da15d609da1ff!std::panicking::continue_panic_fmt+0xa4
0000008c`bc975430 00007ffd`0ca9eedc std_822da15d609da1ff!std::panicking::rust_begin_panic+0x9
0000008c`bc975460 00007ffd`0cab1e7b std_822da15d609da1ff!core::panicking::panic_fmt+0x3c
0000008c`bc9754c0 00007ffd`07177a58 std_822da15d609da1ff!core::str::slice_error_fail+0x2fb
0000008c`bc9755d0 00007ffd`072d1314 rustc_158b3752b6ee1be6!ZN5rustc4lint7builtin22BuiltinLintDiagnostics3run17h4bb3542f622712c3E+0xb38
0000008c`bc9757a0 00007ffd`072dc943 rustc_158b3752b6ee1be6!ZN5rustc4lint7context12EarlyContext8check_id17h7f0064d9b036bf15E+0x4a4
*** WARNING: Unable to verify checksum for rustc_driver-b7c23ebe756beb8b.dll
0000008c`bc975b30 00007ffd`1a82e309 rustc_158b3752b6ee1be6!ZN5rustc4lint7context15check_ast_crate17hf056b9ed4f51c850E+0x3f3
0000008c`bc975e60 00007ffd`1a827b4c rustc_driver_b7c23ebe756beb8b!ZN12rustc_driver6driver11count_nodes17h72939dcc6d63838eE+0x2339
0000008c`bc977c90 00007ffd`1a8c8099 rustc_driver_b7c23ebe756beb8b!ZN12rustc_driver6driver13compile_input17h7b701af89d68eeaeE+0x65c
0000008c`bc97a010 00007ffd`1a82698c rustc_driver_b7c23ebe756beb8b!ZN12rustc_driver22run_compiler_with_pool17h1e7dab3c08524b6dE+0x1009
0000008c`bc97bbf0 00007ffd`1a7db3fb rustc_driver_b7c23ebe756beb8b!ZN6syntax12with_globals17h3b145e5e099dcffdE+0x60c
0000008c`bc97d190 00007ffd`0ca9a732 rustc_driver_b7c23ebe756beb8b!ZN101_$LT$std..panic..AssertUnwindSafe$LT$F$GT$$u20$as$u20$core..ops..function..FnOnce$LT$$LP$$RP$$GT$$GT$9call_once17h66b0d8204f471548E+0xbb
0000008c`bc97f9e0 00007ffd`1a8c43e7 std_822da15d609da1ff!panic_unwind::__rust_maybe_catch_panic+0x22
0000008c`bc97fa40 00007ffd`1a8d664e rustc_driver_b7c23ebe756beb8b!ZN12rustc_driver7profile22profile_queries_thread17heb501a38eb0cafebE+0x5807
*** WARNING: Unable to verify checksum for rustc.exe
0000008c`bc97fd10 00007ff6`3ea71006 rustc_driver_b7c23ebe756beb8b!ZN12rustc_driver4main17h055a7d7a4cdec127E+0xe
0000008c`bc97fd40 00007ffd`0ca68647 rustc!__acrt_rg_language_count <PERF> (rustc+0x1006)
(Inline Function) --------`-------- std_822da15d609da1ff!std::rt::lang_start_internal::{{closure}}+0xc
0000008c`bc97fd70 00007ffd`0ca9a732 std_822da15d609da1ff!std::panicking::try::do_call<closure,i32>+0x17
0000008c`bc97fda0 00007ffd`0ca65133 std_822da15d609da1ff!panic_unwind::__rust_maybe_catch_panic+0x22
(Inline Function) --------`-------- std_822da15d609da1ff!std::panicking::try+0x33
(Inline Function) --------`-------- std_822da15d609da1ff!std::panic::catch_unwind+0x33
0000008c`bc97fe00 00007ff6`3ea7107a std_822da15d609da1ff!std::rt::lang_start_internal+0xa3
0000008c`bc97fea0 00007ff6`3ea71299 rustc!main+0x2a
(Inline Function) --------`-------- rustc!invoke_main+0x22
0000008c`bc97fed0 00007ffd`23a613d2 rustc!__scrt_common_main_seh+0x11d
0000008c`bc97ff10 00007ffd`258754f4 kernel32!BaseThreadInitThunk+0x22
0000008c`bc97ff40 00000000`00000000 ntdll!RtlUserThreadStart+0x34
```

</details>

zackmdavis commented:
Hm, the `snippet.split_at((insertion_span.lo().0 - path_span.lo().0) as usize);` is probably not legitimate in the presence of macros. I'll take a closer look soon.

zackmdavis commented:
I added [a check](https://github.com/rust-lang/rust/blob/53cebfde949f42ca0d406b42520032017e123a35/src/librustc/lint/builtin.rs#L457-L462) that I'm quite confident will avoid the panic we saw in the AppVeyor Windows build. This _does_ imply that there exist some cases with macros for which we'll issue no suggestion (or possibly a wrong suggestion), but I don't think this should be a blocking concern. (Note that the UI test [does demonstrate some simple macro scenarios](https://github.com/rust-lang/rust/blob/53cebfde949f42ca0d406b42520032017e123a35/src/test/ui/in-band-lifetimes/elided-lifetimes.stderr#L37-L53) that we handle correctly.) Thus, I argue that this should be re-r+'d.

@nikomatsakis :checkered_flag:

nikomatsakis commented:
@zackmdavis great!

nikomatsakis commented:
@bors r+ p=1

This is important for EP2

bors commented:
:pushpin: Commit 53cebfde949f42ca0d406b42520032017e123a35 has been approved by `nikomatsakis`

<!-- @bors r=nikomatsakis 53cebfde949f42ca0d406b42520032017e123a35 -->

bors commented:
:hourglass: Testing commit 53cebfde949f42ca0d406b42520032017e123a35 with merge a48cbacac725c0c5c76e4954b7d9ff657d591dab...

bors commented:
:broken_heart: Test failed - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.8262)

kennytm commented on src/librustc/lint/builtin.rs:
```diff
@@ -442,6 +443,36 @@ impl BuiltinLintDiagnostics {
+                    // `Path<T>` expression with `Path<'_, T>`, rather than inserting `'_, `
+                    // at a point (which makes for an ugly/confusing label)
+                    if let Ok(snippet) = sess.codemap().span_to_snippet(path_span) {
+                        let insertion_index = (insertion_span.lo().0 - path_span.lo().0) as usize;
```
```
   Compiling termcolor v0.3.6
thread 'main' panicked at 'attempt to subtract with overflow', librustc\lint\builtin.rs:456:47
note: Run with `RUST_BACKTRACE=1` for a backtrace.
error: Could not compile `serde_derive`.
```

😂

zackmdavis commented:
@kennytm @nikomatsakis I forgot that integer overflow panics (in non-release builds). Should be fixed now.

bors commented:
:hourglass: Testing commit 41d5c0ce1fb4618c98b3ed07d657bdedfc98c959 with merge ffaf3d2567c18abdaaac689776105fc99aaa36a3...

bors commented:
:sunny: Test successful - [status-appveyor](https://ci.appveyor.com/project/rust-lang/rust/build/1.0.8267), [status-travis](https://travis-ci.org/rust-lang/rust/builds/406914502?utm_source=github_status&utm_medium=notification)
Approved by: nikomatsakis
Pushing ffaf3d2567c18abdaaac689776105fc99aaa36a3 to master...

rust-highfive commented:
📣 Toolstate changed by rust-lang/rust#52069!

Tested on commit rust-lang/rust@ffaf3d2567c18abdaaac689776105fc99aaa36a3.
Direct link to PR: <https://github.com/rust-lang/rust/pull/52069>

💔 clippy-driver on windows: test-fail → build-fail (cc @Manishearth @llogiq @mcarton @oli-obk, @rust-lang/infra).
💔 clippy-driver on linux: test-fail → build-fail (cc @Manishearth @llogiq @mcarton @oli-obk, @rust-lang/infra).
💔 rls on windows: test-pass → build-fail (cc @nrc, @rust-lang/infra).
💔 rls on linux: test-pass → build-fail (cc @nrc, @rust-lang/infra).

oli-obk commented:
I think we have `#[deny(warnings)]`
